lightning/src/offers/async_receive_offer_cache.rs

```diff
@@ -304,37 +316,48 @@ impl AsyncReceiveOfferCache {
 	/// until it succeeds, see [`AsyncReceiveOfferCache`] docs.
 	pub(super) fn cache_pending_offer(
 		&mut self, offer: Offer, offer_paths_absolute_expiry_secs: Option<u64>, offer_nonce: Nonce,
-		update_static_invoice_path: Responder, duration_since_epoch: Duration,
-	) -> Result<u16, ()> {
+		update_static_invoice_path: Responder, duration_since_epoch: Duration, cache_slot: u16,
+	) -> Result<(), ()> {
 		self.prune_expired_offers(duration_since_epoch, false);
 
 		if !self.should_build_offer_with_paths(
 			offer.paths(),
 			offer_paths_absolute_expiry_secs,
+			cache_slot,
 			duration_since_epoch,
 		) {
 			return Err(());
 		}
 
-		let idx = match self.needs_new_offer_idx(duration_since_epoch) {
-			Some(idx) => idx,
-			None => return Err(()),
-		};
-
-		match self.offers.get_mut(idx) {
-			Some(offer_opt) => {
-				*offer_opt = Some(AsyncReceiveOffer {
+		match self.offers.get_mut(cache_slot as usize) {
+			Some(slot) => {
+				*slot = Some(AsyncReceiveOffer {
 					offer,
 					created_at: duration_since_epoch,
 					offer_nonce,
 					status: OfferStatus::Pending,
 					update_static_invoice_path,
-				});
+				})
+			},
+			None => {
+				debug_assert!(false, "We checked the cache slot in should_build_offer_with_paths");
```

Comment:
This assert message isn't fully clear to me. We checked the cache slot, and something was wrong with it?